src/protocols/displayport.stanza

```diff
@@ -0,0 +1,220 @@
+
+<DOC>
+#use-added-syntax(jitx)
+defpackage jsl/protocols/displayport:
+  import core
+  import jitx
+  import jitx/commands
+
+  import jsl/bundles
+  import jsl/errors
+  import jsl/si/helpers
+  import jsl/si/couplers
+  import jsl/pin-assignment
+
+
+doc: \<DOC>
+@brief DisplayPort Generation enums
+This is a fixed list of possible generation definitions for DisplayPort
+<DOC>
+
+public pcb-enum jsl/protocols/displayport/DPVersion:
+  DP1p0  ; 10.80 Gbit/s / 4 lanes =>  2.7 Gbit/s / 10 UI =>  270 MHz clk
+  DP1p1  ; 10.80 Gbit/s / 4 lanes =>  2.7 Gbit/s / 10 UI =>  270 MHz clk
+  DP1p2  ; 21.60 Gbit/s / 4 lanes =>  5.4 Gbit/s / 10 UI =>  540 MHz clk
+  DP1p2a ; 21.60 Gbit/s / 4 lanes =>  5.4 Gbit/s / 10 UI =>  540 MHz clk
+  DP1p3  ; 32.40 Gbit/s / 4 lanes =>  8.1 Gbit/s / 10 UI =>  810 MHz clk
+  DP1p4  ; 32.40 Gbit/s / 4 lanes =>  8.1 Gbit/s / 10 UI =>  810 MHz clk
+  DP1p4a ; 32.40 Gbit/s / 4 lanes =>  8.1 Gbit/s / 10 UI =>  810 MHz clk
+  DP2p0  ; 80.00 Gbit/s / 4 lanes => 20.0 Gbit/s / 10 UI => 2000 MHz clk
+  DP2p1  ; 80.00 Gbit/s / 4 lanes => 20.0 Gbit/s / 10 UI => 2000 MHz clk
+  DP2p1a ; 80.00 Gbit/s / 4 lanes => 20.0 Gbit/s / 10 UI => 2000 MHz clk
+
+doc: \<DOC>
+@brief DisplayPort Bundle
+The Data Bundle consists of 4 lane pairs (1 RX or TX pair set = 1 lane)
+The bundle also constains an AUX channel (differential pair) and a HPD pin
+<DOC>
+
+public pcb-bundle displayport :
+  name = "DisplayPort"
+  description = "DisplayPort Digital Display Interface"
+  port aux  : diff-pair
+  port lane : diff-pair[4]
+  port hpd  : pin ; hot plug detect
+
+
+defn constrain-ch (sw:Toleranced, ml:Double, a:JITXObject, b:JITXObject) :
+  skew-window(sw, a => b)
+  max-loss(ml, a => b)
+
+doc: \<DOC>
+@brief Construct a DisplayPort lane.
+
+This function will construct the lane connections of a DisplayPort connection. The `src` and `dst`
+arguments must both be of type `displayport`.
+
+@param sw The intra-pair skew time difference for the differential signals being connected.
+@param ml The maximum loss in dB for all differential signals being connected.
+@param cap Optional DC blocking capacitor to be inserted between the TX and RX pairs. Note that
+there are 2 blocking caps inserted, one on each of the P,N signals of the pair.
+@param src This is assumed to be the "Tx" side of the link. This is expected to be a port of `Bundle` type `displayport`
+@param dst This is assumed to be the "Rx" side of the link. This is expected to be a port of `Bundle` type `displayport`
+
+<DOC>
+public defn connect-displayport-lane (sw:Toleranced, ml:Double, cap:Instantiable|False, src:JITXObject, dst:JITXObject) :
+  inside pcb-module:
+    for i in indices(src.lane) do :
+      match(cap) :
+        (c:Instantiable) :
+          inst bl-cap : dp-coupler(c)
+          require tx1 : dual-pair from bl-cap
+          topo-net(src.lane[i], tx1.A)
+          topo-net(tx1.B, dst.lane[i])
```

Comment:
We should not change the duplication here. 
I don't think we fully understand the abstraction yet. 
There is clearly a repetitive case here that I think has additional features. As we discussed - ESD protection diodes are another case where there maybe a similar pattern. 

Let's think about how this might be better expressed. In the future, if/when we come up with a solution we can refactor.

Reply:
Is there a code change that I need to do for your comment?